src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.cs

```diff
@@ -334,63 +334,16 @@ public void DeleteSubKeyTree(string subkey, bool throwOnMissingSubKey)
 
             subkey = FixupName(subkey); // Fixup multiple slashes to a single slash
 
-            RegistryKey? key = InternalOpenSubKeyWithoutSecurityChecks(subkey, true);
-            if (key != null)
-            {
-                using (key)
-                {
-                    if (key.SubKeyCount > 0)
-                    {
-                        string[] keys = key.GetSubKeyNames();
-
-                        for (int i = 0; i < keys.Length; i++)
-                        {
-                            key.DeleteSubKeyTreeInternal(keys[i]);
-                        }
-                    }
-                }
+            int ret = Interop.Advapi32.RegDeleteTree(_hkey, subkey);
```

Comment:
Doc says:
> A handle to an open registry key. The key must have been opened with the following access rights: DELETE, KEY_ENUMERATE_SUB_KEYS, and KEY_QUERY_VALUE. For more information, see [Registry Key Security and Access Rights](https://learn.microsoft.com/en-us/windows/desktop/SysInfo/registry-key-security-and-access-rights).

but for RegDeleteKeyEx it does not require this
> A handle to an open registry key. The access rights of this key do not affect the delete operation. For more information about access rights, see [Registry Key Security and Access Rights](https://learn.microsoft.com/en-us/windows/desktop/SysInfo/registry-key-security-and-access-rights).

Is there any way this change could cause something to fail that worked before?

Reply:
I don't think it can cause deletion to fail. Should we add a manual access check to restore existing behavior?

Reply:
you mean, you've verified that all codepaths that lead here provide a handle that has DELETE, KEY_ENUMERATE_SUB_KEYS, and KEY_QUERY_VALUE?

Reply:
This is a public API. I mean that we may perform additional access check to match the document.
I've checked Windows source code and `RegDeleteKeyEx` uses the same enumeration process. I'm not sure where the access check happens though.

Reply:
The following code fails to delete with P/Invoke, but successes with `DeleteSubKeyTree` on .NET 7:

```csharp
        static void Main(string[] args)
        {
            using var regtree = Registry.CurrentUser.OpenSubKey("TestDelete", RegistryKeyPermissionCheck.ReadWriteSubTree ,
                RegistryRights.QueryValues | RegistryRights.EnumerateSubKeys);
            regtree.DeleteSubKeyTree("123");
            // int ret = RegDeleteTree(regtree.Handle, "123");
        }

        [DllImport("advapi32.dll", ExactSpelling = true)]
        static extern int RegDeleteTreeW(SafeRegistryHandle handle, string subkeyName);
```

I need better understanding about registry permissions though.

Reply:
So this would imply that it would fail with your change, but would have succeeded in .NET 7.

It would be good to add a test case that covers this scenario as part of your change.

The nuance here is that `RegDeleteKeyEx` opens a new handle for the subkey with delete permission:
> The function opens the subkey with the DELETE access right.

Whereas this is not done for `RegDeleteTree`.  I did notice the following in the docs for `RegDeleteTree`
> The name of the key. This key must be a subkey of the key identified by the hKey parameter. If this parameter is NULL, the subkeys and values of hKey are deleted.

Maybe you could open that subkey handle yourself then specify a NULL name to `RegDeleteTree` to satisfy the handle permission requirements, maintaining compatibility.

Reply:
It seems that the behavior of `RegDeleteTree` itself is inconsistent. In the following code:
```csharp
using var regtree = Registry.CurrentUser.OpenSubKey("TestDelete", RegistryKeyPermissionCheck.ReadWriteSubTree,
                RegistryRights.QueryValues);
var subKey = regtree.OpenSubKey("123", false);
int ret = RegDeleteTreeW(subKey.Handle, "");
```
`RegDeleteTree` returns ACCESS_DENIED, however the subkeys are deleted.

Reply:
A possibility is catching such failure and using "The old way"... not sure that's worth it

Reply:
Perhaps we need a matrix for behaviors of the old and new way. The old way should also cause partial deletion in some cases.

Reply:
Notes after reading more about registry:
- The `RegistryKeyPermissionCheck` or `isWritable` check only happens at managed side. It can be easily replicated.
- Registry checks the access of current user, not desired access of the handle.

So special casing "" should be enough. No matrix needed actually.

Reply:
> The following code fails to delete with P/Invoke, but successes with DeleteSubKeyTree on .NET 7:

I can't reproduce the failure now. Maybe I missed something.

Reply:
The failures should be the documented case of `RegDeleteTree`:

> If the key has values, it must be opened with KEY_SET_VALUE or this function will fail with ERROR_ACCESS_DENIED.

I think covering the case should be enough. ACL handlings shouldn't be changed.